File: web/src/api/cleaning.ts
/**
 * AI Cleaning 数据清洗规则推荐 API
 *
 * API 规范:
 * - v1 版本使用统一的 ApiResponse 格式
 * - 保留旧版 API 以确保向后兼容
 */

import client from './client';
import type { ApiResponse } from './types';

// 重新导出统一类型
export type { ApiResponse };

// ============================================================
// 类型定义
// ============================================================

export interface AnalyzeQualityRequest {
  table_name: string;
  database?: string;
  sample_size?: number;
}

export interface AnalyzeQualityResult {
  table_name: string;
  total_rows: number;
  quality_score: number;
  issues: Array<{
    column: string;
    issue_type: string;
    count: number;
    percentage: number;
  }>;
}

export interface RecommendRulesRequest {
  table_name: string;
  database?: string;
}

export interface CleaningRule {
  id?: string;
  name: string;
  description: string;
  type: string;
  scenario?: string;
  config?: Record<string, unknown>;
}

/** Extended cleaning rule with recommendation metadata */
export interface CleaningRuleRecommendation extends CleaningRule {
  column?: string;
  confidence?: number;
}

export interface RecommendRulesResult {
  table_name: string;
  recommended_rules: CleaningRule[];
}

export interface GenerateConfigRequest {
  table_name: string;
  rules: string[];
  output_format?: 'seatunnel' | 'hop';
}

export interface GenerateConfigResult {
  config: string;
  format: string;
}

// ============================================================
// v1 版本 API（推荐使用，统一响应格式）
// ============================================================

/** 分析数据质量 */
export async function analyzeQualityV1(
  params: AnalyzeQualityRequest
): Promise<ApiResponse<AnalyzeQualityResult>> {
  const resp = await client.post<ApiResponse<AnalyzeQualityResult>>(
    '/api/proxy/cleaning/v1/analyze',
    params
  );
  return resp.data;
}

/** AI 推荐清洗规则 */
export async function recommendRulesV1(
  params: RecommendRulesRequest
): Promise<ApiResponse<RecommendRulesResult>> {
  const resp = await client.post<ApiResponse<RecommendRulesResult>>(
    '/api/proxy/cleaning/v1/recommend',
    params
  );
  return resp.data;
}

/** 获取清洗规则模板 */
export async function getCleaningRulesV1(): Promise<ApiResponse<CleaningRule[]>> {
  const resp = await client.get<ApiResponse<CleaningRule[]>>(
    '/api/proxy/cleaning/v1/rules'
  );
  return resp.data;
}

/** 生成 SeaTunnel Transform 配置 */
export async function generateConfigV1(
  params: GenerateConfigRequest
): Promise<ApiResponse<GenerateConfigResult>> {
  const resp = await client.post<ApiResponse<GenerateConfigResult>>(
    '/api/proxy/cleaning/v1/generate-config',
    params
  );
  return resp.data;
}

// ============================================================
// 旧版 API（向后兼容，逐步废弃）
// ============================================================

/** 分析数据质量 */
export async function analyzeQuality(params: AnalyzeQualityRequest) {
  const resp = await client.post('/api/proxy/cleaning/api/cleaning/analyze', params);
  return resp.data;
}

/** AI 推荐清洗规则 */
export async function recommendRules(params: RecommendRulesRequest) {
  const resp = await client.post('/api/proxy/cleaning/api/cleaning/recommend', params);
  return resp.data;
}

/** 获取清洗规则模板 */
export async function getCleaningRules() {
  const resp = await client.get('/api/proxy/cleaning/api/cleaning/rules');
  return resp.data;
}

/** 生成 SeaTunnel Transform 配置 */
export async function generateConfig(params: GenerateConfigRequest) {
  const resp = await client.post('/api/proxy/cleaning/api/cleaning/generate-config', params);
  return resp.data;
}

File: services/portal/routers/cleaning.py
"""AI Cleaning 代理路由 - AI 清洗推荐"""

import os

import httpx
from fastapi import APIRouter, Depends, Request, Response
from pydantic import BaseModel

from services.common.api_response import ApiResponse, ErrorCode, error, success
from services.common.auth import TokenPayload, get_current_user
from services.portal.config import settings
from services.portal.routers.proxy import proxy_request

# 服务间通信密钥
SERVICE_SECRET = os.environ.get("SERVICE_SECRET", "internal-service-secret-dev-do-not-use-in-prod")

router = APIRouter(prefix="/api/proxy/cleaning", tags=["AI Cleaning"])


# ============================================================
# 请求模型
# ============================================================

class AnalyzeRequest(BaseModel):
    table_name: str
    database: str | None = None
    sample_size: int | None = 100


class RecommendRequest(BaseModel):
    table_name: str
    database: str | None = None


class GenerateConfigRequest(BaseModel):
    table_name: str
    rules: list[str]
    output_format: str | None = "seatunnel"


# ============================================================
# v1 版本 API（推荐使用，统一响应格式）
# ============================================================

@router.post("/v1/analyze", response_model=ApiResponse)
async def analyze_quality_v1(
    request: AnalyzeRequest,
    user: TokenPayload = Depends(get_current_user),
) -> ApiResponse:
    """分析数据质量（v1 API）"""
    try:
        async with httpx.AsyncClient(timeout=60.0) as client:
            resp = await client.post(
                f"{settings.AI_CLEANING_URL}/api/cleaning/analyze",
                json=request.model_dump(),
                headers={"X-Service-Secret": SERVICE_SECRET},
            )
            if resp.status_code == 200:
                return success(data=resp.json())
            return error(
                message=f"AI 清洗服务错误: {resp.status_code}",
                code=ErrorCode.EXTERNAL_SERVICE_ERROR,
            )
    except Exception as e:
        return error(
            message=f"AI 清洗服务异常: {str(e)}",
            code=ErrorCode.EXTERNAL_SERVICE_ERROR,
        )


@router.post("/v1/recommend", response_model=ApiResponse)
async def recommend_rules_v1(
    request: RecommendRequest,
    user: TokenPayload = Depends(get_current_user),
) -> ApiResponse:
    """AI 推荐清洗规则（v1 API）"""
    try:
        async with httpx.AsyncClient(timeout=60.0) as client:
            resp = await client.post(
                f"{settings.AI_CLEANING_URL}/api/cleaning/recommend",
                json=request.model_dump(),
                headers={"X-Service-Secret": SERVICE_SECRET},
            )
            if resp.status_code == 200:
                return success(data=resp.json())
            return error(
                message=f"AI 清洗服务错误: {resp.status_code}",
                code=ErrorCode.EXTERNAL_SERVICE_ERROR,
            )
    except Exception as e:
        return error(
            message=f"AI 清洗服务异常: {str(e)}",
            code=ErrorCode.EXTERNAL_SERVICE_ERROR,
        )


@router.get("/v1/rules", response_model=ApiResponse)
async def get_cleaning_rules_v1(
    user: TokenPayload = Depends(get_current_user),
) -> ApiResponse:
    """获取清洗规则模板（v1 API）"""
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            resp = await client.get(
                f"{settings.AI_CLEANING_URL}/api/cleaning/rules",
                headers={"X-Service-Secret": SERVICE_SECRET},
            )
            if resp.status_code == 200:
                return success(data=resp.json())
            return error(
                message=f"AI 清洗服务错误: {resp.status_code}",
                code=ErrorCode.EXTERNAL_SERVICE_ERROR,
            )
    except Exception as e:
        return error(
            message=f"AI 清洗服务异常: {str(e)}",
            code=ErrorCode.EXTERNAL_SERVICE_ERROR,
        )


@router.post("/v1/generate-config", response_model=ApiResponse)
async def generate_config_v1(
    request: GenerateConfigRequest,
    user: TokenPayload = Depends(get_current_user),
) -> ApiResponse:
    """生成 SeaTunnel Transform 配置（v1 API）"""
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            resp = await client.post(
                f"{settings.AI_CLEANING_URL}/api/cleaning/generate-config",
                json=request.model_dump(),
                headers={"X-Service-Secret": SERVICE_SECRET},
            )
            if resp.status_code == 200:
                return success(data=resp.json())
            return error(
                message=f"AI 清洗服务错误: {resp.status_code}",
                code=ErrorCode.EXTERNAL_SERVICE_ERROR,
            )
    except Exception as e:
        return error(
            message=f"AI 清洗服务异常: {str(e)}",
            code=ErrorCode.EXTERNAL_SERVICE_ERROR,
        )


# ============================================================
# 旧版 API（向后兼容，直接代理）
# ============================================================

@router.api_route("/{path:path}", methods=["GET", "POST", "PUT", "DELETE"])
async def cleaning_proxy(
    path: str,
    request: Request,
    user: TokenPayload = Depends(get_current_user),
) -> Response:
    """代理 AI 清洗服务请求（旧版 API，向后兼容）"""
    return await proxy_request(
        request=request,
        target_base_url=settings.AI_CLEANING_URL,
        target_path=path,
    )
